### ISSUE TITLE
How do I unit test Dapper?

### ISSUE THREAD

**Issue by Almenon** [StackExchange/Dapper#1074]
The Dapper readme has no info on unit testing, and suprisingly there is not much tutorials online either.

The two main unit test libraries I was able to find were:

* DapperWrapper (not updated since 2014!!!)
* Moq.Dapper (requires .net core v2, but many people are still on VS2015+.net 4.6.2 or lesser)

The dapper repo, meanwhile, appears to have integration tests using a actual database.  While having a database for integration tests is ideal, it should also be possible to mock the connection so a database is not required.

**Comment by mgravell** [comment]
Frankly, I'm of the opinion that pure unit tests would be essentially useless here - most of the interesting problems are due to how it *actually* works with real providers, and as such I'm not massively inclined to invest much time trying to make large chunks of it "pure" unit tests. I'm also not sure it would serve a real purpose - the integration tests don't take very long to run.

So: what actual real problem are you trying to solve here? Do you have a specific intent in mind? Having pure tests just for the sake of it isn't actually a useful goal. I kinda don't care much whether I can successfully talk to mocks - I care that I can talk to a database.

Perhaps an in-process database might be worth exploring, but...

**Comment by Almenon** [comment]
Oh, I'm not talking about unit testing dapper itself.  I'm talking about how to unit test in projects that use dapper

**Comment by billrob** [comment]
@Almenon You can wrap the dapper calls with your own interface.  We have IDbExecutor that wraps the functions of dapper that we need.  So at run time the DapperDbExecutor is the concrete used by real code and we mock IDbExecutor for our unit tests.

**Comment by Manirajss** [comment]
@Almenon I have used Sqlite to test queries in my project which written using Dapper. You may have small drawbacks such as sqlite is not supporting schemas and some syntax difference in sql functions.

**Comment by iby-dev** [comment]
I am looking for a dapper inmemory provider implementation - i am looking to move away from traditional pure unit tests just like Marc says. EFCore has some great new features but for dapper news is a little thin on the ground. I don't want to Mock anything, I want to spin up an inmemory db and then fire my DAL code at it and then create some scenarios around it. As Marc says, this more useful to me than the countless thousands of traditional tests I have done over the years. Any one has ideas on this please get in touch.

**Comment by nrjohnstone** [comment]
I agree with Marc that unit tests are not appropriate at this level.

If you are using an architectural style such as ports and adapters, then Dapper is an implementation detail of the particular adapter for a given port, in this case interfaces on the "persistence" port. As such, all of your "unit" testing should be done against an in memory implementation of your persistence interface.

Personally I suggest to people that they should forget about unit testing Dapper, it's a wrapper around database interaction and the only way to test that reliably is with integration tests otherwise all you are potentially testing and assert is strings and calls to mocks, which is next to useless and is binding intimately with your implementation details, which leads to a nightmare when refactoring anything.

The value of most unit tests is against your domain/core logic, keep them there and implement in memory implementations of your persistence interfaces.

For more concrete examples, have a look here 

https://github.com/nrjohnstone/ports-adapters-examples/tree/master/example01

**Comment by codeapologist** [comment]
I have created a light abstraction library to help facilitate unit testing. As previously mentioned, this is not intended to test your queries with Dapper, but to provide a way to mock Dapper in your unit tests.  I plan to keep it maintained and ensure it passes all of Dapper's integration tests. [DataAbstractions.Dapper](https://github.com/codeapologist/DataAbstractions.Dapper)

**Comment by julealgon** [comment]
> Frankly, I'm of the opinion that pure unit tests would be essentially useless here - most of the interesting problems are due to how it actually works with real providers, and as such I'm not massively inclined to invest much time trying to make large chunks of it "pure" unit tests. I'm also not sure it would serve a real purpose - the integration tests don't take very long to run.
>
> So: what actual real problem are you trying to solve here? Do you have a specific intent in mind? Having pure tests just for the sake of it isn't actually a useful goal. I kinda don't care much whether I can successfully talk to mocks - I care that I can talk to a database.

@mgravell while I can see where you are coming from and agree to some extent with the mindset, you should seriously reconsider not forcing it on consumers of the library as some companies have strict rules on unit test coverage and this can lead to severe difficulties when testing with Dapper.

Even if you do not think unit testing is beneficial from your perspective, it would be great if the library was written in a more testable way by exposing a few interfaces on top of the static extension methods. There is no need to get rid of the extension methods of course, this would just be a different way for tapping into the same functionality. Think `AutoMapper` here: it has all the static classes, but it also exposes the `IMapper` interface, which can very easily be mocked.

This allows the consumers of the library to decide if they want to go for full unit test testability (by injecting `IMapper`) or the simpler `Mapper` static class approach and rely on integration/functional tests to validate the behavior.

TL;DR I agree with the issue proposed here regardless of my opinion about the "usefulness" of unit tests. The client should decide how useful it is to them to have those pure unit tests.

**Comment by nrjohnstone** [comment]
That fact that AutoMapper made the bad mistake of putting an interface over something that essentially just copies property A to property B, to me demonstrates how far down the path of crazy a certain view of unit testing is.

The IMapper was not a bad design decision in itself, but what it allows people to do now is couple to the internal details of a Class about the fact that "mapping" takes place and the signatures that are used to do the mapping.

The very fact that a misguided proportion of the "unit" testing crowd believe that coupling a test fixture to every class in your solution is a great thing makes me believe that Dapper should resist doing this because at least it prevents people doing this.

Unit Test != Class Testing

My main reason for posting is I'm passionate about the great micro-orm that Dapper is and don't want to see it providing tools that directly encourage bad practices such as what AutoMapper has done.

**Comment by reckface** [comment]
I **always** recommend and use dapper for data access where inline SQL is already extensively used, and that's because it's a solid proven library. I don't need or want to test dapper.
What would be nice is a way to ensure that the objects/parameters passed to dapper match what the database is expecting. If you've worked with large legacy codebases there's a tendency for stored procedure parameters to change independent of the .net code, and there's no way to find out until runtime.
I need a way to verify the parameter object contains the parameters in the stored procedure metadata, or in the inline query as part of the test suite.

```
var items = connection.Query<Item>("Select Name, Id, Quantity From SomeTable WHERE Name like @name AND Quantity > @quantity", new { Name, Quantity}); 

// the tests
parameterObjectProperties.Should().Contain("name", "quantity");
```

**Comment by 2piix** [comment]
I'm just a lowly data analyst, but I have some testing needs (which are probably closer to integration testing) as well that are very much unmet by Dapper.  (No offense, it's a great project)

Here's my scenario: I need to repeatably run a query to make sure that it always returns known good data (where the known good data is manually pulled out of the live database).  I'm not interested in setting up "scenarios" or "mocking".  I just want to run the query and make sure that, when tested against a live clone of production, the data we pull matches known good data. If we find a bug, we add a test case and work on the query until all the tests pass.  (You know, like "unit testing", but in an RO environment instead of an isolated, fake one.  The unit I'm testing is the actual SQL query.)

So, I found myself writing a lot of C# (not one of my strengths), when I thought, "Surely I can't be the only person who uses Dapper to write queries and tests, there *must* be something out there for me!"  And upon that realization, I decided to search and found that there isn't any "official"

**Comment by MovGP0** [comment]
I implemented a custom base interface for DBContext types: 

```csharp
    public interface IDbContext
    {
        Task<int> SaveChangesAsync(CancellationToken cancellationToken);

        Task<IDbContextTransaction> BeginTransactionAsync(CancellationToken cancellationToken);

        IDbConnection GetDbConnection();

        // define Dapper methods here
        Task<int> ExecuteAsync(string sql, object param = null, IDbTransaction transaction = null, int? commandTimeout = null, CommandType? commandType = null);
    }
```

Then I implement the context like this:

```csharp
    public interface IMyContext : IDbContext
    {
        // ...
    }

    public sealed class MyContext : DbContext, IMyContext
    {
        public MyContext(DbContextOptions<MyContext> options) : base(options)
        {
        }

        public Task<IDbContextTransaction> BeginTransactionAsync(CancellationToken cancellationToken)
        {
            return Database.BeginTransactionAsync(cancellationToken);
        }

        public IDbConnection GetDbConnection()
        {
            return Database.GetDbConnection();
        }

        // wrap Dapper methods
        public Task<int> ExecuteAsync(string sql, object param = null, IDbTransaction transaction = null, int? commandTimeout = null, CommandType? commandType = null)
        {
            return GetDbConnection().ExecuteAsync(sql, param, transaction, commandTimeout, commandType);
        }

        // ...
```
Now i can easily mock everything: 

```csharp
var substitute = Substitute.For<IMyContext>();
```

**Comment by MovGP0** [comment]
> @Almenon I have used Sqlite to test queries in my project 

well, that's technically an **Integration Test** and not an **Unit Test**. Note the difference.

**Comment by michelvd** [comment]
I don't think you have to unit test the Dapper calling code all the time, and of course you don't have to test the internal working of Dapper, but if you have a complex query or one with some parameters, it isn't weird to test if that goes well? Or if a method has 2 queries in it, to unit test that the data returned from query a is properly set to the parameter in query 2? 

I don't think you can assume that all code is tidy and solid and all, some code just isn't and especially then you want to write a unit test.

In these days it strikes me as strange that a library has multiple untestable methods, because they are implemented as extension methods and/or static methods.

Then saying: "you shouldn't be testing any code that is using Dapper' is not the solution imo.

**Comment by JibSail** [comment]
I have a model with some troublesome getters and setters.

It would be nice to be able to test how they behave with Dapper; Considering the value add is preventing breaking code changes in the model. 

Dapper.Moq doesn't seem like it would help in this case.

I guess I could try mocking the actual db connection... I'll let you know how that goes.

**Comment by DanielLidstromTicket** [comment]
You can solve this by customizing the `IDbConnection`. Here's a complete sample that maps the query string to DataSet, effectively simulating database access. All taken care of by a custom `IDbConnection`.

```csharp
void Main()
{
    var dataSets = new Dictionary<string, DataSet>
    {
        // provide queries and the expected results, this is your stub
        {
            "select top 2 * from hotel",
            new List<Hotel>
            {
                new Hotel { Code = "H-123", HotelId = Guid.NewGuid() },
                new Hotel { Code = "H-456", HotelId = Guid.NewGuid() },
            }.ToDataSet()
        },
        {
            "select * from customer",
            new List<Customer>
            {
                new Customer { Id = 1, Name = "Daniel" }
            }.ToDataSet()
        }
    };

    // the connection is the extension point, setup in your tests
    var conn = new MyDbConnection(dataSets);

    // probably your application code starts like this, pass the connection
    // to whatever you use (IoC, etc)
    conn.Open();
    conn.Query<Hotel>("select top 2 * from hotel").Dump();
    conn.Query<Customer>("select * from customer").Dump();
}

// Define other methods and classes here
class Hotel
{
    public string Code { get; set; }
    public Guid HotelId { get; set; }
}

class Customer
{
    public int Id { get; set; }
    public string Name { get; set; }
}

public static class MyExtensions
{
    public static DataSet ToDataSet<T>(this IList<T> list)
    {
        Type elementType = typeof(T);
        DataSet ds = new DataSet();
        DataTable t = new DataTable();
        ds.Tables.Add(t);

        // add a column to table for each public property on T
        foreach (var propInfo in elementType.GetProperties())
        {
            Type colType = Nullable.GetUnderlyingType(propInfo.PropertyType) ?? propInfo.PropertyType;

            t.Columns.Add(propInfo.Name, colType);
        }

        // go through each property on T and add each value to the table
        foreach (T item in list)
        {
            DataRow row = t.NewRow();

            foreach (var propInfo in elementType.GetProperties())
            {
                row[propInfo.Name] = propInfo.GetValue(item, null) ?? DBNull.Value;
            }

            t.Rows.Add(row);
        }

        return ds;
    }
}

private class MyDbConnection : IDbConnection
{
    private readonly Dictionary<string, DataSet> dataSets;

    public MyDbConnection(Dictionary<string, DataSet> dataSets)
    {
        this.dataSets = dataSets;
    }

    public string ConnectionString { get => ""; set => throw new NotImplementedException(); }

    public int ConnectionTimeout => throw new NotImplementedException();

    public string Database => throw new NotImplementedException();

    public ConnectionState State => ConnectionState.Open;

    public IDbTransaction BeginTransaction()
    {
        throw new NotImplementedException();
    }

    public IDbTransaction BeginTransaction(System.Data.IsolationLevel il)
    {
        throw new NotImplementedException();
    }

    public void ChangeDatabase(string databaseName)
    {
        throw new NotImplementedException();
    }

    public void Close()
    {
        throw new NotImplementedException();
    }

    public IDbCommand CreateCommand()
    {
        return new MyCommand(dataSets);
    }

    public void Dispose()
    {
        throw new NotImplementedException();
    }

    public void Open()
    {
    }

    private class MyCommand : IDbCommand
    {
        private readonly Dictionary<string, DataSet> dataSets;
        private DataSet currentDataSet;

        public MyCommand(Dictionary<string, DataSet> dataSets)
        {
            this.dataSets = dataSets;
        }

        public IDbConnection Connection { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }

        public IDbTransaction Transaction { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }

        public string CommandText
        {
            get
            {
                throw new NotImplementedException();
            }

            set
            {
                if (dataSets.TryGetValue(value, out currentDataSet) == false)
                {
                    throw new Exception($"No DataSet configured for query '{value}', update your test setup");
                }
            }
        }

        public int CommandTimeout { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }

        public CommandType CommandType { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }

        public IDataParameterCollection Parameters => throw new NotImplementedException();

        public UpdateRowSource UpdatedRowSource { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }

        public void Cancel()
        {
            throw new NotImplementedException();
        }

        public IDbDataParameter CreateParameter()
        {
            throw new NotImplementedException();
        }

        public void Dispose()
        {
        }

        public int ExecuteNonQuery()
        {
            throw new NotImplementedException();
        }

        public IDataReader ExecuteReader()
        {
            return currentDataSet.CreateDataReader();
        }

        public IDataReader ExecuteReader(CommandBehavior behavior)
        {
            return currentDataSet.CreateDataReader();
        }

        public object ExecuteScalar()
        {
            throw new NotImplementedException();
        }

        public void Prepare()
        {
            throw new NotImplementedException();
        }
    }
}
```

The output:

![image](https://user-images.githubusercontent.com/60345105/79761714-199bf780-8322-11ea-9fb4-cb56887f5738.png)

**Comment by dlidstrom** [comment]
You can solve this by customizing the `IDbConnection`. Here's a complete sample that maps the query string to DataSet, effectively simulating database access. All taken care of by a custom `IDbConnection`.

```csharp
void Main()
{
    var dataSets = new Dictionary<string, DataSet>
    {
        // provide queries and the expected results, this is your stub
        {
            "select top 2 * from hotel",
            new List<Hotel>
            {
                new Hotel { Code = "H-123", HotelId = Guid.NewGuid() },
                new Hotel { Code = "H-456", HotelId = Guid.NewGuid() },
            }.ToDataSet()
        },
        {
            "select * from customer",
            new List<Customer>
            {
                new Customer { Id = 1, Name = "Daniel" }
            }.ToDataSet()
        }
    };

    // the connection is the extension point, setup in your tests
    var conn = new MyDbConnection(dataSets);

    // probably your application code starts like this, pass the connection
    // to whatever you use (IoC, etc)
    conn.Open();
    conn.Query<Hotel>("select top 2 * from hotel").Dump();
    conn.Query<Customer>("select * from customer").Dump();
}

// Define other methods and classes here
class Hotel
{
    public string Code { get; set; }
    public Guid HotelId { get; set; }
}

class Customer
{
    public int Id { get; set; }
    public string Name { get; set; }
}

public static class MyExtensions
{
    public static DataSet ToDataSet<T>(this IList<T> list)
    {
        Type elementType = typeof(T);
        DataSet ds = new DataSet();
        DataTable t = new DataTable();
        ds.Tables.Add(t);

        // add a column to table for each public property on T
        foreach (var propInfo in elementType.GetProperties())
        {
            Type colType = Nullable.GetUnderlyingType(propInfo.PropertyType) ?? propInfo.PropertyType;

            t.Columns.Add(propInfo.Name, colType);
        }

        // go through each property on T and add each value to the table
        foreach (T item in list)
        {
            DataRow row = t.NewRow();

            foreach (var propInfo in elementType.GetProperties())
            {
                row[propInfo.Name] = propInfo.GetValue(item, null) ?? DBNull.Value;
            }

            t.Rows.Add(row);
        }

        return ds;
    }
}

private class MyDbConnection : IDbConnection
{
    private readonly Dictionary<string, DataSet> dataSets;

    public MyDbConnection(Dictionary<string, DataSet> dataSets)
    {
        this.dataSets = dataSets;
    }

    public string ConnectionString { get => ""; set => throw new NotImplementedException(); }

    public int ConnectionTimeout => throw new NotImplementedException();

    public string Database => throw new NotImplementedException();

    public ConnectionState State => ConnectionState.Open;

    public IDbTransaction BeginTransaction()
    {
        throw new NotImplementedException();
    }

    public IDbTransaction BeginTransaction(System.Data.IsolationLevel il)
    {
        throw new NotImplementedException();
    }

    public void ChangeDatabase(string databaseName)
    {
        throw new NotImplementedException();
    }

    public void Close()
    {
        throw new NotImplementedException();
    }

    public IDbCommand CreateCommand()
    {
        return new MyCommand(dataSets);
    }

    public void Dispose()
    {
        throw new NotImplementedException();
    }

    public void Open()
    {
    }

    private class MyCommand : IDbCommand
    {
        private readonly Dictionary<string, DataSet> dataSets;
        private DataSet currentDataSet;

        public MyCommand(Dictionary<string, DataSet> dataSets)
        {
            this.dataSets = dataSets;
        }

        public IDbConnection Connection { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }

        public IDbTransaction Transaction { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }

        public string CommandText
        {
            get
            {
                throw new NotImplementedException();
            }

            set
            {
                if (dataSets.TryGetValue(value, out currentDataSet) == false)
                {
                    throw new Exception($"No DataSet configured for query '{value}', update your test setup");
                }
            }
        }

        public int CommandTimeout { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }

        public CommandType CommandType { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }

        public IDataParameterCollection Parameters => throw new NotImplementedException();

        public UpdateRowSource UpdatedRowSource { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }

        public void Cancel()
        {
            throw new NotImplementedException();
        }

        public IDbDataParameter CreateParameter()
        {
            throw new NotImplementedException();
        }

        public void Dispose()
        {
        }

        public int ExecuteNonQuery()
        {
            throw new NotImplementedException();
        }

        public IDataReader ExecuteReader()
        {
            return currentDataSet.CreateDataReader();
        }

        public IDataReader ExecuteReader(CommandBehavior behavior)
        {
            return currentDataSet.CreateDataReader();
        }

        public object ExecuteScalar()
        {
            throw new NotImplementedException();
        }

        public void Prepare()
        {
            throw new NotImplementedException();
        }
    }
}
```

The output:

![image](https://user-images.githubusercontent.com/60345105/79761714-199bf780-8322-11ea-9fb4-cb56887f5738.png)

**Comment by mmulhearn** [comment]
Unit testing with Dapper is definitely something that needs to be implemented.  My approach tends to be abstracting dependencies out as much as possible so you aren't tied to as many dependencies as possible.  As such, my database access is through a custom `ISqlClient` and Dapper is used in a concrete implementation of `ISqlClient`.  This does make things more complicated and generic, but this would allow us to move from Dapper to another ORM should Dapper discontinue or we find an irreconcilable issue with Dapper.

With this approach, Dapper data splitting is done as generically as possible.  What I'd like to do is mock out the sql connection and a sql command so it returns a mocked data reader that pretends to be the result from the database.  That would allow me to unit test the `map` delegate of `SqlMapper.QueryAsync<T>` and verify that my code is handling the data splits correctly.

The issue is that Dapper moves from interfaces (`IDbConnection` and `IDbCommand`) to unmockable concretes (`DbConnection` and `DbCommand`) so mocking out the connection and command fail as the interfaces don't cast.

I'm still looking around the web for solutions, but there may be none at this point in time.

**Comment by losito105** [comment]
> @Almenon You can wrap the dapper calls with your own interface. We have IDbExecutor that wraps the functions of dapper that we need. So at run time the DapperDbExecutor is the concrete used by real code and we mock IDbExecutor for our unit tests.

We do something similar on my project w/ different implementations of a shared database accessor interface for each of our services. These can easily be mocked, and their methods stubbed using standard Moq. In line queries are another story though...

**Comment by epicbugs** [comment]
I have found a way which works without any additional layers. 

```C#

// Arrange
Mock<DbConnection> dbConnection = new Mock<DbConnection>();
Mock<DbCommand> dbCommand = new Mock<DbCommand>();
Mock<IDataParameterCollection> dbCommandParams = new Mock<IDataParameterCollection>();

dbCommand.As<IDbCommand>().SetupAllProperties();
dbCommand.As<IDbCommand>().SetupGet(c => c.Parameters).Returns(dbCommandParams.Object);
dbCommand.As<IDbCommand>().Setup(c => c.CreateParameter()).Returns(() => new Mock<IDbDataParameter>().SetupAllProperties().Object);
dbConnection.Protected().Setup<DbCommand>("CreateDbCommand").Returns(dbCommand.Object);

// Act
var sql = "UPDATE some SET column = @value WHERE id = @id";
var parms = new { value = "hello", id = Guid.NewGuid() };
await dbConnection.Object.ExecuteAsync(sql, parms);

// Assert
dbCommand.Object.CommandText.Trim().ShouldBe(sql);
dbCommandParams.Verify(p => p.Add(Moq.It.Is<IDbDataParameter>(param =>
    param.ParameterName == "id" && ((Guid)param.Value == parms.id)
)), Times.Once());

```

It is not suitable for all scenarios but with modifications the general idea should be usable elsewhere as well.

**Comment by nrjohnstone** [comment]
This is the insane type of unit testing that should be prevented by anyone who has any kind of sense. Unit tests written in this manner couple your unit testing code DIRECTLY with internal implementation details. The only way to effectively test database connections and SQL is integration testing against something simple like a docker container using test containers.

Sure there might be some mappers and other logic in your database/dapper layer that you MIGHT want to unit test directly, but the best way is to integration test from the outer layer (if you are using a repository pattern this would be your FooRepository implementation in Dapper) and then after invoking the method on your SUT, go and check the database for the expected result.

Testing in this manner (that also includes 'unit' testing, when 'unit' are the group of classes within a boundary fulfilling a unit of functionality) means that your tests are not directly coupled with all sorts of internal gunk like method signatures etc... 

You will feel the weight of this "mockist" unit testing if you ever have to do large scale refactoring, or introducing functionality changes, as the thousands of tests that "mockist" approaches encourage get in your way with breaking changes everywhere.

A fundamental software quality metric is coupling, and between boundaries you WANT low coupling. That includes your unit tests.

Unit testing != Class testing != Mock testing

One caveat where I might consider not using the real implementation is if I am testing failure conditions such as throwing exceptions on timeouts... and even then I would NOT use a mocking library. It is much easier  to just implement an in memory version of your abstraction (IFooRepository -> FooRepositoryInMemory) and give your self new public methods that are NOT on the interface that your test fixture can use. This avoids the method signature insanity that mocks encourage since you now have your own specific methods that the test fixture can use to add/read data or force exceptions to be thrown.

**Comment by nrjohnstone** [comment]
> > [@Almenon](https://github.com/Almenon) I have used Sqlite to test queries in my project
> 
> well, that's technically an **Integration Test** and not an **Unit Test**. Note the difference.

A unit test is NOT mocking every single damn class in the ctor. A unit test is a unit of functionality within a boundary, this can easily involve N number of REAL classes that do not have to be mocked at all.

The only things that need stubbing are abstractions on the boundary of your assembly, the classic being the abstractions that represent your persistence layer... eg IFooRepository 

If you are adding interfaces to every class JUST so you can create mocks, you need to evaluate what level of engineering you are operating at.

**Comment by epicbugs** [comment]
> This is the insane type of unit testing that should be prevented by anyone who has any kind of sense. Unit tests written in this manner couple your unit testing code DIRECTLY with internal implementation details. The only way to effectively test database connections and SQL is integration testing against something simple like a docker container using test containers.
> 
> Sure there might be some mappers and other logic in your database/dapper layer that you MIGHT want to unit test directly, but the best way is to integration test from the outer layer (if you are using a repository pattern this would be your FooRepository implementation in Dapper) and then after invoking the method on your SUT, go and check the database for the expected result.

I am glad that at least some of us have the luxury of refactoring any legacy system they don't like before delivering any useful features. Unfortunately not everyone is so blessed, and not every legacy system is using the repository design pattern.

As for the "insane type", I think the problem is objectively a lot smaller than you imagine it to be. Dapper is relatively stable, and even if it changes the implementation too much, then a legacy system doesn't really have any direct need to adopt a newer version. Or if it absolutely must and the owners hired you to refactor it then you will simply have the opportunity to present a larger invoice.

In principle I agree that a repository pattern etc is preferable, but I take exception to the idealism it has to enforced everywhere no matter how long it takes or how much it costs.

Coming back to the unit test prototype I shared - it has drawbacks (both objective and subjective) but it works as intended.

**Comment by nrjohnstone** [comment]
> > This is the insane type of unit testing that should be prevented by anyone who has any kind of sense. Unit tests written in this manner couple your unit testing code DIRECTLY with internal implementation details. The only way to effectively test database connections and SQL is integration testing against something simple like a docker container using test containers.
> > Sure there might be some mappers and other logic in your database/dapper layer that you MIGHT want to unit test directly, but the best way is to integration test from the outer layer (if you are using a repository pattern this would be your FooRepository implementation in Dapper) and then after invoking the method on your SUT, go and check the database for the expected result.
> 
> I am glad that at least some of us have the luxury of refactoring any legacy system they don't like before delivering any useful features. Unfortunately not everyone is so blessed, and not every legacy system is using the repository design pattern.
> 
> As for the "insane type", I think the problem is objectively a lot smaller than you imagine it to be. Dapper is relatively stable, and even if it changes the implementation too much, then a legacy system doesn't really have any direct need to adopt a newer version. Or if it absolutely must and the owners hired you to refactor it then you will simply have the opportunity to present a larger invoice.
> 
> In principle I agree that a repository pattern etc is preferable, but I take exception to the idealism it has to enforced everywhere no matter how long it takes or how much it costs.
> 
> Coming back to the unit test prototype I shared - it has drawbacks (both objective and subjective) but it works as intended.

But do you actually understand what this type of testing does to your overall application? When you do this to every class in your application that had interfaces (or you add interfaces to every class just so you can use mocks), do you understand how your test code is now directly coupled to internal implementation details of every class in that assembly being tested ?

This is not something you want to inflict on any meaningful solution that needs to be maintained and have features developed. Usually what slows down features being added is the 1,000,000 "unit tests with mocks coupled to every single method signature.

My reason for posting here is that even if you can't grok this, hopefully people following this thread will do some research and understand that you can achieve excellent test coverage (unit and integration) without needing to couple to every single class.

Also, if your existing system (I won't use legacy here because every system is legacy as soon as it is in prod) does not easily allow you add the new feature and you don't refactor your architecture to match the new requirements, you're not even doing agile, since the whole point of agile is to avoid excessive "up front" design and do the design when it matters. If you just smoosh your feature in without adjusting your design (yes 'refactoring') then nothing at all can save your or that business.

**Comment by epicbugs** [comment]
> But do you actually understand what this type of testing does to your overall application? When you do this to every class in your application that had interfaces (or you add interfaces to every class just so you can use mocks), do you understand how your test code is now directly coupled to internal implementation details of every class in that assembly being tested ?

Dear John - first of all, I am not sure how to explain it any better to you but the legacy system which inspired this solution is **not** my application, neither in terms of legal ownership nor code lines. Secondly, unit testing as such does not do anything per se to any application. Tests are there just to tell you if something has broken in some expected way. Nothing less, nothing more.

When it comes to coupling then every piece of code which calls another piece of code introduces a degree of coupling. The only question is whether a particular degree of coupling in a particular piece of code is an acceptable tradeoff considering all the other relevant factors. I understand the implications and for my particular case the tradeoffs are acceptable to me. I am not saying they have to be acceptable for you. I am perfectly okay with you or anyone else never using this approach.

> This is not something you want to inflict on any meaningful solution that needs to be maintained and have features developed. Usually what slows down features being added is the 1,000,000 "unit tests with mocks coupled to every single method signature.

I think you are exaggerating and extrapolating without considering any concrete situation.

> My reason for posting here is that even if you can't grok this, hopefully people following this thread will do some research and understand that you can achieve excellent test coverage (unit and integration) without needing to couple to every single class.

I am pretty sure the engineers who bother to read this ticket are aware of it and so am I. The authors of Dapper have had every opportunity to make their library mockable but so far they haven't. Eventually the community takes things into their own hands and as you can see it is not exactly pretty but that is what happens when reasonable feature requests are not respected.

> Also, if your existing system (I won't use legacy here because every system is legacy as soon as it is in prod) does not easily allow you add the new feature and you don't refactor your architecture to match the new requirements, you're not even doing agile, since the whole point of agile is to avoid excessive "up front" design and do the design when it matters. If you just smoosh your feature in without adjusting your design (yes 'refactoring') then nothing at all can save your or that business.

I disagree with your opinion that every system is legacy. I think the main reason you are attempting to discharge the meaning of that term is because you want to avoid recognizing the reality that legacy systems can have a very different cost-benefit calculus compared to modern systems. Refactoring might be useful but engineers usually aren't deciding on time allocation.

**Comment by julealgon** [comment]
> This is the insane type of unit testing that should be prevented by anyone who has any kind of sense.
> 
> A unit test is NOT mocking every single damn class in the ctor. A unit test is a unit of functionality within a boundary, this can easily involve N number of REAL classes that do not have to be mocked at all.

@nrjohnstone what people define as a "unit" is up to them, so your attempt at objectively dismissing what a unit test could be is just flawed and shortsighted. A unit test that mocks every single dependency in its constructor can _definitely_ be a 100% valid unit test. You may opt not to have those types of tests in _your_ codebase for whatever reason _you_ judge valid, but that doesn't mean you should try to pass your subjective understanding to others as an absolute. Lastly, you assume that people writing these kinds of unit tests don't also have the logic covered in an integration test: that's a bad assumption.

**Comment by AdaskoTheBeAsT** [comment]
my approach is to write integration tests which uses TestContainers with given real database - mocking db is in my opinion bad idea

it is ultra easy with SqlServer because you can create sql server project and create deployment schema script - then only fill db with test data and thats it - in other dbs you need little more effort - still doable
repositories - integ tests
services and other classes - unit
controllers - integ with JasperFX Alba
win

i use definition - if something can be tested completely alone - unit
if it requires part of external system - queue, db, azure services whatever - integ
